src/analysis/mean_inequalities.lean

```diff
@@ -103,47 +108,76 @@ begin
   exact this hw
 end
 
-/-- Young's inequality, `ℝ≥0` version -/
-theorem nnreal.young_inequality (a b : ℝ≥0) {p q : ℝ≥0} (hp : 1 < p) (hq : 1 < q)
-  (hpq : 1/p + 1/q = 1) : a * b ≤ a^(p:ℝ) / p + b^(q:ℝ) / q :=
-begin
-  have := nnreal.am_gm2_weighted (1/p) (1/q) (a^(p:ℝ)) (b^(q:ℝ)) hpq,
-  simp only [← nnreal.rpow_mul, one_div_eq_inv, nnreal.coe_div, nnreal.coe_one] at this,
-  rw [mul_inv_cancel, mul_inv_cancel, nnreal.rpow_one, nnreal.rpow_one] at this,
-  { ring at ⊢ this,
-    convert this;
-    { rw [nnreal.div_def, nnreal.div_def], ring } },
-  { exact ne_of_gt (lt_trans zero_lt_one hq) },
-  { exact ne_of_gt (lt_trans zero_lt_one hp) }
-end
+end nnreal
 
-/-- Young's inequality, `ℝ` version -/
-theorem real.young_inequality {a b : ℝ} (ha : 0 ≤ a) (hb : 0 ≤ b)
-  {p q : ℝ} (hp : 1 < p) (hq : 1 < q) (hpq : 1/p + 1/q = 1) :
-  a * b ≤ a^p / p + b^q / q :=
-@nnreal.young_inequality ⟨a, ha⟩ ⟨b, hb⟩ ⟨p, le_trans zero_le_one (le_of_lt hp)⟩
-  ⟨q, le_trans zero_le_one (le_of_lt hq)⟩ hp hq (nnreal.coe_eq.1 hpq)
+namespace real
+
+theorem am_gm2_weighted {w₁ w₂ p₁ p₂ : ℝ} (hw₁ : 0 ≤ w₁) (hw₂ : 0 ≤ w₂)
+  (hp₁ : 0 ≤ p₁) (hp₂ : 0 ≤ p₂) (hw : w₁ + w₂ = 1) :
+  p₁ ^ w₁ * p₂ ^ w₂ ≤ w₁ * p₁ + w₂ * p₂ :=
+nnreal.am_gm2_weighted ⟨w₁, hw₁⟩ ⟨w₂, hw₂⟩ ⟨p₁, hp₁⟩ ⟨p₂, hp₂⟩ $ nnreal.coe_eq.1 $ by assumption
 
-theorem real.pow_am_le_am_pow (w z : ι → ℝ) (hw : ∀ i ∈ s, 0 ≤ w i)
+/-- Young's inequality, a version for nonnegative real numbers. -/
+theorem young_inequality_of_nonneg {a b p q : ℝ} (ha : 0 ≤ a) (hb : 0 ≤ b)
+  (hpq : p.is_conjugate_exponent q) :
+  a * b ≤ a^p / p + b^q / q :=
+by simpa [← rpow_mul, ha, hb, hpq.ne_zero, hpq.symm.ne_zero, div_eq_inv_mul]
+  using real.am_gm2_weighted (le_of_lt hpq.one_div_pos) (le_of_lt hpq.symm.one_div_pos)
```

Comment:
Myabe add `one_div_nonneg` to `conjugate_exponents`, and use it here instead of `le_of_lt (...one_div_pos)`, twice.

Reply:
Sometimes I think about adding `has_lt.lt.le` as an alias for `le_of_lt`.

Reply:
That's a good idea!